byterun/floats.c

```diff
@@ -66,6 +76,57 @@ CAMLexport void caml_Store_double_val(value val, double dbl)
 
 #endif
 
+/*
+ OCaml runtime itself doesn't call setlocale, i.e. it is using
+ standard "C" locale by default, but it is possible that
+ thirt-party code loaded into process does.
+*/
+#ifdef HAS_LOCALE
+
+#ifdef _MSC_VER
+
+#ifndef locale_t
+#define locale_t _locale_t
+#endif
+
+#ifndef freelocale
+#define freelocale _free_locale
+#endif
+
+#ifndef strtod_l
+#define strtod_l _strtod_l
+#endif
+
+#endif
+
+locale_t caml_locale = 0;
+
+#endif
+
+void caml_init_locale(void)
+{
+#ifdef HAS_LOCALE
+  if (0 == caml_locale)
+  {
+#ifdef _MSC_VER
+    _configthreadlocale(_ENABLE_PER_THREAD_LOCALE);
+    caml_locale = _create_locale(LC_NUMERIC, "C");
+#else
+    caml_locale = duplocale(LC_GLOBAL_LOCALE);
+    caml_locale = newlocale(LC_NUMERIC_MASK,"C",caml_locale);
```

Comment:
This will leak memory on OSX (and presumably the BSDs), whose `duplocale`/`newlocale` API is much more reasonable than Linux's.